traversal/selector/exploreUnion.go

```diff
@@ -0,0 +1,95 @@
+package selector
+
+import (
+	"fmt"
+
+	ipld "github.com/ipld/go-ipld-prime"
+)
+
+// ExploreUnion allows selection to continue with two or more distinct selectors
+// while exploring the same tree of data.
+//
+// ExploreUnion can be used to apply a Matcher on one node (causing it to
+// be considered part of a (possibly labelled) result set), while simultaneously
+// continuing to explore deeper parts of the tree with another selector,
+// for example.
+type ExploreUnion struct {
+	Members []Selector
+}
+
+// Interests for ExploreUnion is:
+// - nil (aka all) if any member selector has nil interests
+// - the union of values returned by all member selectors otherwise
+func (s ExploreUnion) Interests() []PathSegment {
+	// Check for any high-cardinality selectors first; if so, shortcircuit.
+	//  (n.b. we're assuming the 'Interests' method is cheap here.)
+	for _, m := range s.Members {
+		if m.Interests() == nil {
+			return nil
+		}
+	}
+	// Accumulate the whitelist of interesting path segments.
+	// TODO: Dedup?
```

Comment:
I'd go a step farther and say they probably ought to be sorted.  Which makes the deduping easier; also, though, because if we're then seeking over a large map that's sorted, we'll tend to get some wins thusly.

I'm fine with merging this without that, though.

Reply:
Yea I'd like to put dedup on the TODO list.